cmd/gpu_plugin/rm/gpu_plugin_resource_manager_test.go

```diff
@@ -0,0 +1,261 @@
+		Spec: v1.PodSpec{
+			Containers: []v1.Container{
+				{
+					Resources: v1.ResourceRequirements{
+						Requests: v1.ResourceList{
+							"gpu.intel.com/i915": resource.MustParse("1"),
+						},
+					},
+				},
+			},
+		},
+	}
+	unAnnotatedTestPod := *properTestPod.DeepCopy()
+	unAnnotatedTestPod.ObjectMeta.Annotations = nil
+	properTestPod2 := *properTestPod.DeepCopy()
+	properTestPod2.ObjectMeta.Name = "TestPod2"
+
+	timeStampedProperTestPod := *properTestPod.DeepCopy()
+	timeStampedProperTestPod.ObjectMeta.Annotations[gasTSAnnotation] = "2"
+
+	timeStampedProperTestPod2 := *properTestPod2.DeepCopy()
+	timeStampedProperTestPod2.ObjectMeta.Annotations[gasTSAnnotation] = "1"
+
+	properContainerRequests := []*v1beta1.ContainerAllocateRequest{{DevicesIDs: []string{"card0-0"}}}
+
+	testCases := []testCase{
+		{
+			name:              "Wrong number of container requests should fail",
+			pods:              []v1.Pod{properTestPod},
+			containerRequests: []*v1beta1.ContainerAllocateRequest{},
+			expectErr:         true,
+		},
+		{
+			name:              "Request for monitor resource should fail",
+			pods:              []v1.Pod{properTestPod},
+			containerRequests: []*v1beta1.ContainerAllocateRequest{{DevicesIDs: []string{"all"}}},
+			expectErr:         true,
+		},
+		{
+			name:              "Zero pending pods should fail",
+			pods:              []v1.Pod{},
+			containerRequests: properContainerRequests,
+			expectErr:         true,
+		},
+		{
+			name:              "Single pending pod without annotations should fail",
+			pods:              []v1.Pod{unAnnotatedTestPod},
+			containerRequests: properContainerRequests,
+			expectErr:         true,
+		},
+		{
+			name:              "Single pending pod should succeed",
+			pods:              []v1.Pod{properTestPod},
+			containerRequests: properContainerRequests,
+			expectErr:         false,
+		},
+		{
+			name:              "Two pending pods without timestamps should fail",
+			pods:              []v1.Pod{properTestPod, properTestPod2},
+			containerRequests: properContainerRequests,
+			expectErr:         true,
+		},
+		{
+			name:              "Two pending pods with timestamps should reduce to one candidate and succeed",
+			pods:              []v1.Pod{timeStampedProperTestPod, timeStampedProperTestPod2},
+			containerRequests: properContainerRequests,
+			expectErr:         false,
+		},
+	}
+
+	for _, tCase := range testCases {
+		rm := newMockResourceManager(tCase)
+		resp, err := rm.ReallocateWithFractionalResources(&v1beta1.AllocateRequest{
+			ContainerRequests: tCase.containerRequests,
+		})
+
+		if (err != nil) && !tCase.expectErr {
+			t.Errorf("test %v unexpected failure, err:%v", tCase.name, err)
+		}
+		if err == nil {
+			if tCase.expectErr {
+				t.Errorf("test %v unexpected success", tCase.name)
+			} else {
+				checkAllocateResponse(resp, t, tCase.name)
+			}
+		}
+	}
+}
+
+func checkAllocateResponse(response *v1beta1.AllocateResponse, t *testing.T, testName string) {
```

Comment:
This looks like a sequence of invariant checks. It may be cleaner with a helper function like
```go
func expectTrue(predicate bool, t *testing.T, testName, format string, args ...interface{}) {
        if !predicate {
                t.Helper()
                t.Errorf(fmt.Sprintf("in test %q ", testName)+format, args...)
        }
}
```
Then you can have a sequence of
```go
...
expectTrue(len(resp.ContainerResponses) == 1, t, tCase.name, "wrong number of container response")
...
```
instead of `checkAllocateResponse()` (which IIUC was introduced to avoid cyclomatic complexity).

Reply:
that was neat, done